lredis/protocol.lua

```diff
@@ -36,43 +39,81 @@ local function encode_inline(arg)
 	return table.concat(arg, " ", 1, n+1)
 end
 
+-- Encode and send a redis command.
+local function send_command(file, arg)
+	local req, err_code = encode_request(arg)
+	if not req then
+		return nil, err_code
+	end
+	req, err_code = file:write(req)
+	if not req then
+		return nil, err_code
+	end
+	req, err_code = file:flush()
+	if not req then
+		return nil, err_code
+	end
+
+	return true
+end
+
 -- Parse a redis response
 local function read_response(file, new_status, new_error, string_null, array_null)
-	local line = assert(file:read("*l"))
-	assert(line:sub(-1, -1) == "\r", "invalid line ending")
+	local line, err_code = file:read("*l")
+	if not line then
+		return nil, err_code or "EOF reached"
+	elseif line:sub(-1, -1) ~= "\r" then
+		return nil, "invalid line ending"
+	end
+
 	local status, data = line:sub(1, 1), line:sub(2, -2)
 	if status == "+" then
 		return new_status(data)
 	elseif status == "-" then
 		return new_error(data)
 	elseif status == ":" then
-		return assert(tonumber(data, 10), "invalid integer")
+		return tonumber(data, 10), "invalid integer"
 	elseif status == "$" then
-		local len = assert(tonumber(data, 10), "invalid bulk string length")
-		if len == -1 then
+		local len = tonumber(data, 10)
+		if not len then
+			return nil, "invalid bulk string length"
+		elseif len == -1 then
 			return string_null
 		elseif len > 512*1024*1024 then -- max 512 MB
-			error("bulk string too large")
+			return nil, "bulk string too large"
```

Comment:
may need to invent error codes?

Reply:
Why?  In what way is this any different to the original code?  After error handling takes place safely at a higher level the result it `error()`'d back to the caller.  There is no change to what the caller receives as a result of this change.